staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -986,6 +1024,56 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 	return allErrs
 }
 
+func extractMaxElements(schema *apiextensions.JSONSchemaProps) *uint64 {
+	switch schema.Type {
+	case "object":
+		if schema.AdditionalProperties != nil {
+			if schema.MaxProperties != nil {
+				maxProps := uint64(*schema.MaxProperties)
+				return &maxProps
+			}
+			return nil
+		}
+		// return 1, and not nil, because even if we're
+		// not looking at a map here, we can assume the object still preserves
+		// the cardinality of its parents
+		cardOne := uint64(1)
+		return &cardOne
+	case "array":
+		if schema.MaxItems != nil {
+			maxItems := uint64(*schema.MaxItems)
+			return &maxItems
+		}
+		return nil
+	default:
+		return nil
+	}
+}
+
+// multiplyWithOverflowGuard returns the product of baseCost and cardinality unless that product
+// would exceed math.MaxUint, in which case math.MaxUint is returned.
+func multiplyWithOverflowGuard(baseCost, cardinality uint64) *uint64 {
+	var resultPtr *uint64
+	var result uint64
+	if baseCost == 0 {
+		// an empty rule can return 0, so guard for that here
+		result = 0
+	} else if math.MaxUint/baseCost < cardinality {
+		result = math.MaxUint
```

Comment:
did we wwant nil or MaxUInt here?

Reply:
I think MaxUint may be better here since we call this function from `getExpressionCost`, and we directly compare the return value of `getExpressionCost` with our cost limit. `MaxUint` will be greater than the limit in all practical cases, which I think is the sort of behavior we want. If we returned nil, we'd have to add a nil check in `getExpressionCost` and then raise a validation error, which I think does the same thing as returning `MaxUInt`.